### PR TITLE
use csv format instead of mkString

### PR DESCRIPTION
**Pull request by abendada**

### CODE REVIEW

**Comment by abendada** [comment]
It is possible to add the following small changes in the package.scala$saveAsCsvFile:
- move the csvFormat outside the mapPartitions loop
- use this csvFormat when generating the header instead of the hardcoded mkString
Thanks in advance

**Comment by falaki** [comment]
@abendada Would you please look into the test failure?

**Comment by abendada** [comment]
It was scalaStyle issue.
I removed the comment (added by my self) causing the issue.
should i add a new pull request
Thanks

File: src/main/scala/com/databricks/spark/csv/CsvRelation.scala [hunk 1]
```diff
@@ -35,7 +35,7 @@ case class CsvRelation protected[spark] (
     useHeader: Boolean,
     delimiter: Char,
     quote: Char,
-    escape: Char,
+    escape: Character,
```
Comment (falaki) [inline src/main/scala/com/databricks/spark/csv/CsvRelation.scala]: Let's keep this a ```Char```

File: src/main/scala/com/databricks/spark/csv/package.scala [hunk 2]
```diff
@@ -30,7 +30,7 @@ package object csv {
                 useHeader: Boolean = true,
                 delimiter: Char = ',',
                 quote: Char = '"',
-                escape: Char = '\\',
+                escape: Character = null,
```
Comment (falaki) [inline src/main/scala/com/databricks/spark/csv/package.scala]: Let's keep the type ```Char```

File: src/main/scala/com/databricks/spark/csv/package.scala [hunk 3]
```diff
@@ -87,36 +89,41 @@ package object csv {
...
+
+      val csvFormat = quoteChar match {
+        case Some(c) => csvFormatBase.withQuote(c)
+        case _ => csvFormatBase
+      }
+
       val generateHeader = parameters.getOrElse("header", "false").toBoolean
       val header = if (generateHeader) {
-        dataFrame.columns.map(c => s""""$c"""").mkString(delimiter)
+        csvFormat.format(dataFrame.columns.map(_.asInstanceOf[AnyRef]):_*)
```
Comment (falaki) [inline src/main/scala/com/databricks/spark/csv/package.scala]: ```_.asInstanceOf[AnyRef]): _*)```

File: src/main/scala/com/databricks/spark/csv/package.scala [hunk 4]
```diff
@@ -87,36 +89,41 @@ package object csv {
...
-        val csvFormat = quoteChar match {
-          case Some(c) => csvFormatBase.withQuote(c)
-          case _ => csvFormatBase
-        }
-
         new Iterator[String] {
           var firstRow: Boolean = generateHeader
 
-          override def hasNext = iter.hasNext
+          override def hasNext = iter.hasNext || firstRow
```
Comment (falaki) [inline src/main/scala/com/databricks/spark/csv/package.scala]: What if iterator is empty?

Comment (falaki) [inline src/main/scala/com/databricks/spark/csv/package.scala]: I see why you added it. Would you please leave a comment as to why. Future contributors will have an easier time understanding it.

File: src/main/scala/com/databricks/spark/csv/CsvRelation.scala (hunk 1 above)
Comment (abendada) [inline src/main/scala/com/databricks/spark/csv/CsvRelation.scala]: It important to be able to set escape to null
just add the following line "thing a ""/"" and a ""\""" to escape.csv and try to assembly
09:09:30.630 ERROR com.databricks.spark.csv.CsvRelation$: Exception while parsing line: "thing a ""/"" and a ""\""". 
java.io.IOException: (startline 1) EOF reached before encapsulated token finished
	at org.apache.commons.csv.Lexer.parseEncapsulatedToken(Lexer.java:282)
	at org.apache.commons.csv.Lexer.nextToken(Lexer.java:152)
	at org.apache.commons.csv.CSVParser.nextRecord(CSVParser.java:498)
	at org.apache.commons.csv.CSVParser.getRecords(CSVParser.java:365)
	at com.databricks.spark.csv.CsvRelation$$anonfun$com$databricks$spark$csv$CsvRelation$$parseCSV$1.apply(CsvRelation.scala:133)
	at com.databricks.spark.csv.CsvRelation$$anonfun$com$databricks$spark$csv$CsvRelation$$parseCSV$1.apply(CsvRelation.scala:130)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:396)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
	at org.apache.spark.sql.execution.Aggregate$$anonfun$execute$1$$anonfun$6.apply(Aggregate.scala:127)
	at org.apache.spark.sql.execution.Aggregate$$anonfun$execute$1$$anonfun$6.apply(Aggregate.scala:124)
	at org.apache.spark.rdd.RDD$$anonfun$14.apply(RDD.scala:634)
	at org.apache.spark.rdd.RDD$$anonfun$14.apply(RDD.scala:634)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:68)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
	at org.apache.spark.scheduler.Task.run(Task.scala:64)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:203)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

Comment (falaki) [inline src/main/scala/com/databricks/spark/csv/CsvRelation.scala]: I see. Then we can change all these types to Character (for sanity).

**Comment by falaki** [comment]
Thanks @abendada. This looks good. Would be good to add a line comment for documentation.

File: src/main/scala/com/databricks/spark/csv/CsvRelation.scala (hunk 1 above)
Comment (abendada) [inline src/main/scala/com/databricks/spark/csv/CsvRelation.scala]: Yes it make sense.
I also notice that the escape character only rendered when quote is set to null

**Pull request merged**